File: libs/fluxc-tests/src/test/java/org/wordpress/android/fluxc/theme/ThemeStoreUnitTest.java
package org.wordpress.android.fluxc.theme;

import android.content.Context;

import com.yarolegovich.wellsql.WellSql;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mockito;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;
import org.wordpress.android.fluxc.Dispatcher;
import org.wordpress.android.fluxc.TestSiteSqlUtils;
import org.wordpress.android.fluxc.model.SiteModel;
import org.wordpress.android.fluxc.model.ThemeModel;
import org.wordpress.android.fluxc.network.rest.wpcom.theme.ThemeRestClient;
import org.wordpress.android.fluxc.persistence.SiteSqlUtils;
import org.wordpress.android.fluxc.persistence.ThemeSqlUtils;
import org.wordpress.android.fluxc.persistence.WellSqlConfig;
import org.wordpress.android.fluxc.site.SiteUtils;
import org.wordpress.android.fluxc.store.ThemeStore;

import java.util.ArrayList;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;

@RunWith(RobolectricTestRunner.class)
public class ThemeStoreUnitTest {
    private final ThemeStore mThemeStore = new ThemeStore(new Dispatcher(), Mockito.mock(ThemeRestClient.class));

    @Before
    public void setUp() {
        Context appContext = RuntimeEnvironment.getApplication().getApplicationContext();
        WellSqlConfig config = new WellSqlConfig(appContext);
        WellSql.init(config);
        config.reset();
    }

    @Test
    public void testActiveTheme() throws SiteSqlUtils.DuplicateSiteException {
        final SiteModel site = SiteUtils.generateWPComSite();
        TestSiteSqlUtils.INSTANCE.getSiteSqlUtils().insertOrUpdateSite(site);
        assertEquals(0, ThemeSqlUtils.getActiveThemeForSite(site).size());

        final ThemeModel firstTheme = generateTestTheme(site.getId(), "first-active", "First Active");
        final ThemeModel secondTheme = generateTestTheme(site.getId(), "second-active", "Second Active");
        firstTheme.setActive(true);
        secondTheme.setActive(true);

        // set first theme active and verify
        mThemeStore.setActiveThemeForSite(site, firstTheme);
        List<ThemeModel> activeThemes = ThemeSqlUtils.getActiveThemeForSite(site);
        assertNotNull(activeThemes);
        assertEquals(1, activeThemes.size());
        ThemeModel firstActiveTheme = activeThemes.get(0);
        assertEquals(firstTheme.getThemeId(), firstActiveTheme.getThemeId());
        assertEquals(firstTheme.getName(), firstActiveTheme.getName());

        // set second theme active and verify
        mThemeStore.setActiveThemeForSite(site, secondTheme);
        activeThemes = ThemeSqlUtils.getActiveThemeForSite(site);
        assertNotNull(activeThemes);
        assertEquals(1, activeThemes.size());
        ThemeModel secondActiveTheme = activeThemes.get(0);
        assertEquals(secondTheme.getThemeId(), secondActiveTheme.getThemeId());
        assertEquals(secondTheme.getName(), secondActiveTheme.getName());
    }

    @Test
    public void testInsertOrUpdateTheme() throws SiteSqlUtils.DuplicateSiteException {
        final SiteModel site = SiteUtils.generateJetpackSiteOverRestOnly();
        TestSiteSqlUtils.INSTANCE.getSiteSqlUtils().insertOrUpdateSite(site);

        final String testThemeId = "fluxc-ftw";
        final String testThemeName = "FluxC FTW";
        final String testUpdatedName = testThemeName + " v2";
        final ThemeModel insertTheme = generateTestTheme(site.getId(), testThemeId, testThemeName);

        // verify theme doesn't already exist
        assertNull(mThemeStore.getInstalledThemeByThemeId(site, testThemeId));

        // insert new theme and verify it exists
        ThemeSqlUtils.insertOrUpdateSiteTheme(site, insertTheme);
        ThemeModel insertedTheme = mThemeStore.getInstalledThemeByThemeId(site, testThemeId);
        assertNotNull(insertedTheme);
        assertEquals(testThemeName, insertedTheme.getName());

        // update the theme and verify the updated attributes
        insertedTheme.setName(testUpdatedName);
        ThemeSqlUtils.insertOrUpdateSiteTheme(site, insertedTheme);
        insertedTheme = mThemeStore.getInstalledThemeByThemeId(site, testThemeId);
        assertNotNull(insertedTheme);
        assertEquals(testUpdatedName, insertedTheme.getName());
    }

    @Test
    public void testInsertOrReplaceWpThemes() {
        final List<ThemeModel> firstTestThemes = generateThemesTestList(20);
        final List<ThemeModel> secondTestThemes = generateThemesTestList(30);
        final List<ThemeModel> thirdTestThemes = generateThemesTestList(10);

        // first add 20 themes and make sure the count is correct
        ThemeSqlUtils.insertOrReplaceWpComThemes(firstTestThemes);
        assertEquals(20, mThemeStore.getWpComThemes(ids(firstTestThemes)).size());

        // next add a larger list of themes (with 20 being duplicates) and make sure the count is correct
        ThemeSqlUtils.insertOrReplaceWpComThemes(secondTestThemes);
        assertEquals(30, mThemeStore.getWpComThemes(ids(secondTestThemes)).size());

        // lastly add a smaller list of themes (all duplicates) and make sure count is correct
        ThemeSqlUtils.insertOrReplaceWpComThemes(thirdTestThemes);
        assertEquals(10, mThemeStore.getWpComThemes(ids(thirdTestThemes)).size());
    }

    @Test
    public void testRemoveThemesWithNoSite() {
        final List<ThemeModel> testThemes = generateThemesTestList(20);
        final List<String> themeIds = ids(testThemes);

        // insert and verify count
        assertEquals(0, mThemeStore.getWpComThemes(themeIds).size());
        ThemeSqlUtils.insertOrReplaceWpComThemes(testThemes);
        assertEquals(testThemes.size(), mThemeStore.getWpComThemes(themeIds).size());

        // remove and verify count
        ThemeSqlUtils.removeWpComThemes();
        assertEquals(0, mThemeStore.getWpComThemes(themeIds).size());
    }

    private ThemeModel generateTestTheme(int siteId, String themeId, String themeName) {
        @SuppressWarnings("deprecation") ThemeModel theme = new ThemeModel();
        theme.setLocalSiteId(siteId);
        theme.setThemeId(themeId);
        theme.setName(themeName);
        return theme;
    }

    private List<ThemeModel> generateThemesTestList(int num) {
        List<ThemeModel> testThemes = new ArrayList<>();
        for (int i = 0; i < num; ++i) {
            testThemes.add(generateTestTheme(0, "themeid" + i, "themename" + i));
        }
        return testThemes;
    }

    private List<String> ids(List<ThemeModel> themes) {
        List<String> themeIds = new ArrayList<>();
        for (ThemeModel theme : themes) {
            themeIds.add(theme.getThemeId());
        }
        return themeIds;
    }
}

File: libs/fluxc-tests/src/test/java/org/wordpress/android/fluxc/theme/ThemeStoreUnitTest.kt
package org.wordpress.android.fluxc.theme

import androidx.room.Room
import androidx.test.platform.app.InstrumentationRegistry
import org.assertj.core.api.Assertions.assertThat
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito
import org.robolectric.RobolectricTestRunner
import org.wordpress.android.fluxc.Dispatcher
import org.wordpress.android.fluxc.model.LocalOrRemoteId.LocalId
import org.wordpress.android.fluxc.model.ThemeModel
import org.wordpress.android.fluxc.network.rest.wpcom.theme.ThemeRestClient
import org.wordpress.android.fluxc.persistence.WPAndroidDatabase
import org.wordpress.android.fluxc.site.SiteUtils
import org.wordpress.android.fluxc.store.ThemeStore
import java.io.IOException
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertNull

@RunWith(RobolectricTestRunner::class)
class ThemeStoreUnitTest {
    private lateinit var database: WPAndroidDatabase
    private lateinit var themeStore: ThemeStore

    @Before
    fun setUp() {
        val context = InstrumentationRegistry.getInstrumentation().context
        database = Room.inMemoryDatabaseBuilder(
            context, WPAndroidDatabase::class.java
        ).allowMainThreadQueries().build()

        themeStore = ThemeStore(
            Dispatcher(),
            Mockito.mock(ThemeRestClient::class.java),
            database
        )
    }

    @After
    @Throws(IOException::class)
    fun closeDb() {
        database.close()
    }

    @Test
    fun `when setting active theme, then previous theme is deactivated and new theme is activated`() {
        val site = SiteUtils.generateWPComSite()
        val firstTheme = generateTestTheme(site.id, "first-active", "First Active")
        val secondTheme = generateTestTheme(site.id, "second-active", "Second Active")

        // Set first theme active and verify
        themeStore.setActiveThemeForSite(site, firstTheme)
        val firstActiveTheme = themeStore.getInstalledThemeByThemeId(site, firstTheme.themeId)
        assertNotNull(firstActiveTheme)
        assertEquals(firstTheme.themeId, firstActiveTheme.themeId)
        assertEquals(firstTheme.name, firstActiveTheme.name)
        assertThat(firstActiveTheme.active).isTrue

        // Set second theme active and verify
        themeStore.setActiveThemeForSite(site, secondTheme)
        val secondActiveTheme = themeStore.getInstalledThemeByThemeId(site, secondTheme.themeId)
        assertNotNull(secondActiveTheme)
        assertEquals(secondTheme.themeId, secondActiveTheme.themeId)
        assertEquals(secondTheme.name, secondActiveTheme.name)
        assertThat(secondActiveTheme.active).isTrue

        // Verify first theme is no longer active
        val deactivatedFirstTheme = themeStore.getInstalledThemeByThemeId(site, firstTheme.themeId)
        assertNotNull(deactivatedFirstTheme)
        assertThat(deactivatedFirstTheme.active).isFalse
    }

    @Test
    fun `when setting active theme for site, then theme can be retrieved by theme id`() {
        val testThemeId = "fluxc-ftw"
        val testThemeName = "FluxC FTW"
        val testThemes = listOf(generateTestTheme(0, testThemeId, testThemeName))

        // Verify theme doesn't exist initially
        assertNull(themeStore.getWpComThemeByThemeId(testThemeId))

        // Insert theme via setActiveThemeForSite
        val site = SiteUtils.generateWPComSite()
        themeStore.setActiveThemeForSite(site, testThemes[0])

        // Verify we can retrieve it
        val retrievedTheme = themeStore.getInstalledThemeByThemeId(site, testThemeId)
        assertNotNull(retrievedTheme)
        assertEquals(testThemeName, retrievedTheme.name)
    }

    private fun generateTestTheme(siteId: Int, themeId: String, themeName: String) = ThemeModel(
        siteId = LocalId(siteId),
        themeId = themeId,
        name = themeName,
        demoUrl = null,
        active = false,
        isWpComTheme = false
    )
}
